### PR TITLE
Add endpoint circuit breaker

### PR DESCRIPTION
**Pull request by jirevwe**

### CODE REVIEW

File: pkg/circuit_breaker/circuit_breaker.go [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+		return "half-open"
+	case StateOpen:
+		return "open"
+	default:
+		return fmt.Sprintf("unknown state: %d", s)
+	}
+}
+
+// CircuitBreaker represents a circuit breaker
+type CircuitBreaker struct {
```
Comment (olamilekan000) [inline pkg/circuit_breaker/circuit_breaker.go:162]: I think adding a mutex to the struct will ensure thread safety during concurrent access and modification of values.

File: pkg/circuit_breaker/circuit_breaker.go [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+func (b *CircuitBreaker) String() (s string, err error) {
+	bytes, err := msgpack.EncodeMsgPack(b)
+	if err != nil {
+		return "", err
+	}
+
+	return string(bytes), nil
+}
+
+func (b *CircuitBreaker) tripCircuitBreaker(resetTime time.Time) {
```
Comment (olamilekan000) [inline pkg/circuit_breaker/circuit_breaker.go:183]: you could include

```
	b.mu.Lock()
	defer b.mu.Unlock()
```
while modifying the state of the CB

File: pkg/circuit_breaker/circuit_breaker.go [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+
+func NewCircuitBreakerManager(client redis.UniversalClient) *CircuitBreakerManager {
+	defaultConfig := &CircuitBreakerConfig{
+		SampleRate:                  30,
+		ErrorTimeout:                30,
+		FailureThreshold:            0.1,
+		FailureCount:                10,
+		SuccessThreshold:            5,
+		ObservabilityWindow:         5,
+		NotificationThresholds:      []uint64{5, 10},
```
Comment (olamilekan000) [inline pkg/circuit_breaker/circuit_breaker.go:219]: is there any reason why the config values are hardcoded?

File: pkg/circuit_breaker/circuit_breaker.go (hunk 2 above)
Comment (jirevwe) [reply pkg/circuit_breaker/circuit_breaker.go:183]: We're implementing leader election so only one instance will be running this code. Hence, there's no need to use a mutex

File: pkg/circuit_breaker/circuit_breaker.go (hunk 1 above)
Comment (jirevwe) [reply pkg/circuit_breaker/circuit_breaker.go:162]: We're implementing leader election so only one instance will run this code. Hence, there's no need to use a mutex

File: pkg/circuit_breaker/circuit_breaker.go (hunk 2 above)
Comment (olamilekan000) [reply pkg/circuit_breaker/circuit_breaker.go:183]: I understand that leader election ensures only one instance of the code will run at a time. However, this doesn't eliminate the need for a mutex to guard against potential concurrency issues within that single instance.

File: datastore/models.go [hunk 4]
```diff
@@ -1312,17 +1324,45 @@ type Organisation struct {
...
-	IsSignupEnabled    bool                          `json:"is_signup_enabled" db:"is_signup_enabled"`
-	StoragePolicy      *StoragePolicyConfiguration   `json:"storage_policy" db:"storage_policy"`
-	RetentionPolicy    *RetentionPolicyConfiguration `json:"retention_policy" db:"retention_policy"`
+	UID                string `json:"uid" db:"id"`
+	IsAnalyticsEnabled bool   `json:"is_analytics_enabled" db:"is_analytics_enabled"`
+	IsSignupEnabled    bool   `json:"is_signup_enabled" db:"is_signup_enabled"`
+
+	StoragePolicy        *StoragePolicyConfiguration   `json:"storage_policy" db:"storage_policy"`
+	RetentionPolicy      *RetentionPolicyConfiguration `json:"retention_policy" db:"retention_policy"`
+	CircuitBreakerConfig *CircuitBreakerConfig         `json:"breaker_config" db:"breaker_config"`
```
Comment (mekilis) [inline datastore/models.go:1333]: The name `breaker_config` is not consistent with `circuit_breaker` in config.go

File: sql/1724236900.sql [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+alter table convoy.configurations add column if not exists cb_success_threshold int not null default 1; -- percentage
+alter table convoy.configurations add column if not exists cb_observability_window int not null default 30; -- minutes
+alter table convoy.configurations add column if not exists cb_minimum_request_count int not null default 10;
+alter table convoy.configurations add column if not exists cb_notification_thresholds int[] not null default ARRAY[10, 30, 50];
+alter table convoy.configurations add column if not exists cb_consecutive_failure_threshold int not null default 10;
+create index if not exists idx_delivery_attempts_created_at on convoy.delivery_attempts (created_at);
+create index if not exists idx_delivery_attempts_event_delivery_id_created_at on convoy.delivery_attempts (event_delivery_id, created_at);
+create index if not exists idx_delivery_attempts_event_delivery_id on convoy.delivery_attempts (event_delivery_id);
+
+-- +migrate Down
```
Comment (mekilis) [inline sql/1724236900.sql:15]: The indexes can be dropped, right?

File: cmd/worker/worker.go [hunk 6]
```diff
@@ -327,7 +352,7 @@ func StartWorker(ctx context.Context, a *cli.App, cfg config.Configuration, inte
 
 	// start worker
 	consumer.Start()
-	fmt.Println("Starting Convoy Consumer Pool")
+	lo.Println("Starting Convoy Consumer Pool")
```
[the remark this answers is not in the document]
Comment (jirevwe) [reply cmd/worker/worker.go:355]: Yeah

File: cmd/agent/agent.go [hunk 7]
```diff
@@ -181,7 +177,7 @@ func startServerComponent(ctx context.Context, a *cli.App) error {
 		return nil
 	}
 
-	fmt.Printf("Starting Convoy Agent on port %v\n", cfg.Server.HTTP.AgentPort)
+	log.Println("Starting Convoy Agent on port %v\n", cfg.Server.HTTP.AgentPort)
```
[the remark this answers is not in the document]
Comment (jirevwe) [reply cmd/agent/agent.go:180]: No, I'll change it

File: sql/1724236900.sql (hunk 5 above)
Comment (jirevwe) [reply sql/1724236900.sql:15]: Yeah, I'll add it

File: datastore/models.go (hunk 4 above)
Comment (jirevwe) [reply datastore/models.go:1333]: True, I'll update it

File: pkg/circuit_breaker/circuit_breaker_collector.go [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+	)
+	circuitBreakerConsecutiveFailures = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "", "consecutive_failures"),
+		"Number of consecutive failures for the circuit breaker",
+		[]string{"key"}, nil,
+	)
+	circuitBreakerNotificationsSent = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "", "notifications_sent"),
+		"Number of notifications sent by the circuit breaker",
+		[]string{"key"}, nil,
```
Comment (mekilis) [inline pkg/circuit_breaker/circuit_breaker_collector.go:56]: Is `key` the only required filter for now?

Comment (jirevwe) [reply pkg/circuit_breaker/circuit_breaker_collector.go:56]: Yeah, though, I think at least add the project_id for tenant segmentaton

Comment (mekilis) [reply pkg/circuit_breaker/circuit_breaker_collector.go:56]: Ok, are we adding later or now?

Comment (jirevwe) [reply pkg/circuit_breaker/circuit_breaker_collector.go:56]: I'll do it now, I'll update it in the next commit

File: pkg/clock/clock.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+type realTimeClock struct{}
+
+func (_ *realTimeClock) Now() time.Time { return time.Now() }
+
+// A SimulatedClock is a concrete Clock implementation that doesn't "tick" on its own.
+// Time is advanced by explicit call to the AdvanceTime() or SetTime() functions.
+// This object is concurrency safe.
+type SimulatedClock struct {
+	mu *sync.Mutex
+	t  time.Time // guarded by mu
```
Comment (subomi) [inline pkg/clock/clock.go:31]: @jirevwe, do you think this comment is necessary? It's a more intuitive golang.

Comment (jirevwe) [reply pkg/clock/clock.go:31]: Prob not, lmao

File: pkg/circuit_breaker/store.go [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+	"github.com/go-redsync/redsync/v4"
+	"github.com/go-redsync/redsync/v4/redis/goredis/v9"
+	"github.com/redis/go-redis/v9"
+	"strings"
+	"sync"
+	"time"
+)
+
+type CircuitBreakerStore interface {
+	Lock(ctx context.Context, lockKey string) (*redsync.Mutex, error)
```
Comment (subomi) [inline pkg/circuit_breaker/store.go:17]: @jirevwe Should we couple the store implementation with the distributed lock?

Comment (jirevwe) [reply pkg/circuit_breaker/store.go:17]: I don't think it's a bad idea for now

Comment (subomi) [reply pkg/circuit_breaker/store.go:17]: Fair. It was only a nitpick. :)

File: pkg/circuit_breaker/circuit_breaker.go [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+
+func (b *CircuitBreaker) toHalfOpen() {
+	b.State = StateHalfOpen
+	if b.logger != nil {
+		b.logger.Infof("[circuit breaker] circuit breaker transitioned from open to half-open")
+		b.logger.Debugf("[circuit breaker] circuit breaker state: %+v", b.asKeyValue())
+	}
+}
+
+func (b *CircuitBreaker) ResetCircuitBreaker(resetTime time.Time) {
```
Comment (danvixent) [inline pkg/circuit_breaker/circuit_breaker.go:93]: I think this can just be called Reset()

**Review by subomi — APPROVED** [review]

**Review by danvixent — APPROVED** [review]

**Pull request merged**